src/JsonApiDotNetCore/Errors/JsonApiException.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using JsonApiDotNetCore.Internal;
+using JsonApiDotNetCore.Serialization.Objects;
+
+namespace JsonApiDotNetCore.Errors
+{
+    public class JsonApiException : Exception
+    {
+        private readonly ErrorCollection _errors = new ErrorCollection();
+
+        public JsonApiException(ErrorCollection errorCollection)
+        {
+            _errors = errorCollection;
+        }
+
+        public JsonApiException(Error error, Exception innerException = null)
+            : base(null, innerException)
+        {
+            _errors.Add(error);
+            if (innerException != null)
+            {
+                if (innerException is JsonApiException)
+                {
+                    var innerErrors = ((JsonApiException)innerException)._errors;
+                    foreach (var innerError in innerErrors.Errors)
+                    {
+                        _errors.Add(innerError);
+                    }
+                }
+                
+                _errors.Add(new Error((System.Net.HttpStatusCode) GetStatusCode())
+                {
+                    Source = GetMeta((JsonApiException)innerException)
+                });
+            }
+        }
+
+        public JsonApiException(IEnumerable<Error> errors, Exception innerException = null)
+            : base(null, innerException)
+        {
+            foreach (var error in errors)
+            {
+                _errors.Add(error);
+            }
+
+            if (innerException != null)
+            {
+                _errors.Add(new Error((HttpStatusCode)GetStatusCode())
+                {
+                    Source = GetMeta(innerException)
+                });
+            }
+        }
+
+        // public JsonApiException(Error error)
```

Comment:
What's up with the commented code?

Reply:
I've been a bit sloppy since the eventual aim is to remove this entire code base. The commented out code is unsupported in v4. I'll delete it in the AM